bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareAPIHelper.java

```diff
@@ -0,0 +1,491 @@
+                    decimal = hostname.hashCode();
+                    logger.debug("current hostname: {}, device id: {}", hostname, decimal);
+                }
+            } else {
+                String hostname = localHostAddress.getHostName();
+                decimal = hostname.hashCode();
+                logger.debug("current hostname: {}, device id: {}", hostname, decimal);
+            }
+        } catch (SocketException e) {
+            logger.debug("Socket Exception: {}", e.getMessage());
+        }
+        return decimal;
+    }
+
+    /**
+     * Sets a set of required HTTP headers for the JSON API calls.
+     *
+     * @param request the HTTP connection
+     * @throws ProtocolException
+     */
+    private void setConnectionHeaders(Request request) throws ProtocolException {
+        // headers
+        request.header(HttpHeader.ACCEPT, "application/json, text/plain, */*");
+        request.header(HttpHeader.ACCEPT_ENCODING, "gzip, deflate");
+        request.header(HttpHeader.AUTHORIZATION, "Bearer " + authenticationToken);
+        request.header(HttpHeader.CONNECTION, "keep-alive");
+        request.header(HttpHeader.CONTENT_TYPE, "application/json; utf-8");
+        request.header(HttpHeader.USER_AGENT, API_USER_AGENT);
+        request.header(HttpHeader.REFERER, "https://surepetcare.io/");
+        request.header("Origin", "https://surepetcare.io");
+        request.header("Referer", "https://surepetcare.io");
+        request.header("X-Requested-With", "com.sureflap.surepetcare");
+    }
+
+    /**
+     * Return the "data" element of the API result as a JsonElement.
+     *
+     * @param url The URL of the API call.
+     * @return The "data" element of the API result.
+     * @throws SurePetcareApiException
+     */
+    private JsonElement getDataFromApi(String url) throws SurePetcareApiException {
+        String apiResult = getResultFromApi(url);
+        JsonParser parser = new JsonParser();
+        JsonObject object = (JsonObject) parser.parse(apiResult);
+        return object.get("data");
+    }
+
+    /**
+     * Sends a given object as a JSON payload to the API.
+     *
+     * @param url           the URL
+     * @param requestMethod the request method (POST, PUT etc.)
+     * @param payload       an object used for the payload
+     * @throws SurePetcareApiException
+     */
+    private void setDataThroughApi(String url, HttpMethod method, Object payload) throws SurePetcareApiException {
+        String jsonPayload = SurePetcareConstants.GSON.toJson(payload);
+        postDataThroughAPI(url, method, jsonPayload);
+    }
+
+    /**
+     * Returns the result of a GET API call as a string.
+     *
+     * @param url the URL
+     * @return a JSON string with the API result
+     * @throws SurePetcareApiException
+     */
+    private String getResultFromApi(String url) throws SurePetcareApiException {
+        boolean success = false;
+        String responseData = "";
+        while (!success) {
```

Comment:
I don't like the possibility of this looping endlessly. Can you add a retry limit?

Reply:
Good point. I've changed the code to retry 3 times only.